### DISCUSSION TITLE
Links as wires

### DISCUSSION THREAD

**Discussion by mLuby** [FakeFishGames/Barotrauma#6961] [Ideas and suggestions]
# [Proposal] Links as wires
**TL;DR: links are currently _like_ wires, but can only be modified in the Sub Editor. Let players interact with links _as_ wires, specifically by adding `links_in` and `links_out` pins to the appropriate connection panels and auto-wiring them for compatibility.**

Note: this proposal does _not_ address submarines' shuttle links, outposts' door/hatch/junction box links, or inter-hull links.

### Wire/Unwire in game or in Sub Editor
When a player connects a wire to a links_in and links_out pin, the game will add a link from the links_out machine to the links_in machine. Likewise, if a wire is unwired from either pin, the game unlinks the two newly disconnected machines.

If a link wire is connected through a Relay, shutting off the Relay will unlink the machines. Turning on the Relay will re-link the machines.

### Add pins for links_in and links_out
Since link lines are green in the Sub Editor, the `links_in` and `links_out` pins will have green pin labels (to distinguish them from red power and blue signal) and will use green wires by default. This should eventually be done in XML, but for now the pins can be added in at runtime to ease compatibility.

**Note: Most oxygen generators will need to chain relays to be able to link all the vents.** The alternative is having the `links_in/out` pins support any number of connections—see appendix.

Here are the existing inter-machine links, with the proposed new pins (which can be added at load time, but should eventually be included in the Content XML files).

| Oxygen Generators[¹]() | Vents[²]() | Relay | Ammo Loaders[²]() | Turrets |
|--|--|--|--|--|
| ![Oxygen Generator](https://user-images.githubusercontent.com/2483420/129433971-54cbf24c-34d8-4d48-a1e9-8a0c14fd8978.png) | ![Vent](https://user-images.githubusercontent.com/2483420/129432449-304f4c00-aa70-42d0-b259-d82734a716b0.png) | ![Relay](https://user-images.githubusercontent.com/2483420/129434193-82f54e82-9f12-462d-bc64-f149f78f00b5.png) | ![Ammo Loader](https://user-images.githubusercontent.com/2483420/129432453-83ae7e5f-963d-4436-8cff-49397b334575.png) | ![Turret](https://user-images.githubusercontent.com/2483420/129431971-8ab69089-bec9-423b-a750-f0d24d385cc6.png) |
| **De/constructors** | **Cabinets[²]()/Lockers[²]()** | **Navigation Terminal** | **Status Monitor** | **Sonar Display[³]()** |
| ![De/constructor](https://user-images.githubusercontent.com/2483420/129432439-6bc8c639-849e-4b9d-971e-466beaf0c2dc.png) | ![Cabinet/Locker](https://user-images.githubusercontent.com/2483420/129432074-49d0da6a-cf67-4182-a164-129f04963959.png) | ![Navigation Terminal Connection Panel](https://user-images.githubusercontent.com/2483420/129431692-82abeea6-6185-40d1-84be-0ee430484400.png) | ![Status Monitor](https://user-images.githubusercontent.com/2483420/129432383-edeb75ad-d149-4dae-a385-6f03c9622ebe.png) | ![Sonar Display](https://user-images.githubusercontent.com/2483420/129434367-918418cb-8df8-4bf6-984f-bc9156991cbd.png) |

[¹]() Oxygen Generators only get one `links_out` pin, though using multiple was considered.

[²]() Machine currently doesn't have a Connection Panel.

[³]() Accessing the nav terminal _via_ the status monitor is probably unintentional (or else the same would be true of accessing the Sonar Monitor), so this would fix that bug.

### Auto-wire existing links at load time
**TL;DR: a small forward-compatible code change to wire up Subs and Outposts without player intervention.**

#### Auto-wire pseudo-code
```
onLoad (sub or outpost):
  …
  if not sub.alreadyWiredLinks:
    # Add links_in/out_pins to the machines.
    forEach machine in sub.machines:
      if setOfMachinesWithALinksInPin.has(machine.type):
        addPin(LinksInPin, machine)
      if setOfMachinesWithALinksOutPin.has(machine.type):
        addPin(LinksOutPin, machine)
    # Add wires between links_in/out_pins of linked machines link.
    forEach fromMachine in sub.machines:
      forEach toMachine in fromMachine.linkedTo:
        if (pinFull(fromMachine.links_out_pin or toMachine.links_in_pin)): # Add a relay².
        else: connectWire(fromMachine.links_out_pin, toMachine.links_in_pin, new Wire(green, spriteDepth: 0.999¹))
    sub.alreadyWiredLinks = true
```
[¹]() Wire is at sprite depth 0.999 so it'll be behind everything in game and isn't visible as it runs directly from machine to machine. Another option is to make it transparent.

[²]() To allow relays to control link wires, accessing a linked machine needs to check if the linked machine is a relay. If it is, and it's enabled, the search branches down the relay's `links_out` wires.

### Pros
- Uses existing items, systems, and conventions; no new assets or asset changes required.
- Players who are familiar with wiring will understand linking.
- Enables new player-generated content like toggle-able vents, circuit-switched ammunition changes, and possibly even [utility hardpoints](https://github.com/Regalis11/Barotrauma/discussions/6180).
- Mid-round performance should be nearly identical to current links system, since links only change when unwired or relay-toggled.
- With the [autowiring](#Auto-wire-existing-links-at-load-time), should be a non-breaking change (though not backward compatible).

### Cons
- Auto-link-wiring the machines will slightly increase load time for subs and outposts.
- Strangely-colored player-interactable wires of SpriteDepth .999 (and maybe transparent too) to avoid crazy spiderwebs.

Regarding >5 wires per links pin: to dynamically support large outposts and large custom subs would require way too many `links_out` pins on Oxygen Generators (see [Vanilla submarine vent counts](#Vanilla-submarine-vent-counts)). The dynamic linking could be accomplished by adding chained relays to support any number of links, but that makes the auto-wire links code significantly more complex. For example, where to actually place the relays? What if they spawn under something and are thus unreachable? It's much more straightforward to allow >5 wires only for links pins. However doing so breaks both player and UI assumptions that pins only ever have 0-5 wires attached. Since a key benefit of this proposal _is_ having relay-toggleable links, **it's better to chain relays rather than allow >5 wires.** Relay/link wire placement can be mitigated by manually updating the vanilla subs (which I volunteer to do).

# Appendix

As outposts are dynamically generated, they use the function [`LinkOxygenGenerators`](https://github.com/Regalis11/Barotrauma/search?q=void+LinkOxygenGenerators) to connect the outpost's oxygen generator(s) with vents in the various Outpost Modules (AKA rooms or hulls).

Relays [ReceiveSignal](https://github.com/Regalis11/Barotrauma/blob/d0b1ac96a7ff3a99e9760a1359dd8f4428d7c80f/Barotrauma/BarotraumaShared/SharedSource/Items/Components/Signal/RelayComponent.cs#L172)s here.

Here's an [InitializeLinks](https://github.com/Regalis11/Barotrauma/blob/265a2e7ab3268589ab1ecdcf095eb9ff2fdef20c/Barotrauma/BarotraumaShared/SharedSource/Items/Components/Signal/ConnectionPanel.cs#L85) function.

### Vanilla submarine vent counts

| Class          | Machine                    | Consumption | Amount      | Vents | Image |
|----------------|----------------------------|-------------|-------------|-------|-------|
| Azimuth        | 1 oxygenerator             | 1000W       |  400 O₂     | 10    | [<img alt=Azimuth width=250 src="https://barotraumagame.com/baro-wiki/images/e/e5/Azimuth.png" />](https://barotraumagame.com/baro-wiki/images/e/e5/Azimuth.png) |
| Berilia        | 1 oxygenerator             | 1000W       | 1600 O₂     | 25    | [<img alt=Berilia width=250 src="https://user-images.githubusercontent.com/2483420/137970982-ab171de9-f2f0-44f8-b550-66c947aaf673.png" />](https://barotraumagame.com/baro-wiki/images/3/3f/Berilia.png) |
| Dugong         | 1 oxygenerator             | 1000W       |  400 O₂     |  9    | [<img alt=Dugong width=250 src="https://barotraumagame.com/baro-wiki/images/6/6c/Dugong.png" />](https://barotraumagame.com/baro-wiki/images/6/6c/Dugong.png) |
| Humpback[¹]()  | 1 oxygenerator             | 1000W       |  400 O₂     |  9    | [<img alt=Humpback width=250 src="https://barotraumagame.com/baro-wiki/images/9/94/Humpback.png" />](https://barotraumagame.com/baro-wiki/images/9/94/Humpback.png) |
| Kastrull       | 1 oxygenerator             | 1000W       |  600 O₂     | 17    | [<img alt=Kastrull width=250 src="https://user-images.githubusercontent.com/2483420/137971058-1bb20506-afea-420c-9293-e82a52c7ce30.png" />](https://barotraumagame.com/baro-wiki/images/e/ed/Kastrull.png) |
| Orca           | 1 oxygenerator             | 1000W       |  400 O₂     | 10    | [<img alt=Orca width=250 src="https://barotraumagame.com/baro-wiki/images/7/75/Orca.png" />](https://barotraumagame.com/baro-wiki/images/7/75/Orca.png) |
| R-29           | 1 oxygenerator             | 1000W       |  400 O₂     | 15    | [<img alt="R-29" width=250 src="https://barotraumagame.com/baro-wiki/images/c/c7/R-29.png" />](https://barotraumagame.com/baro-wiki/images/c/c7/R-29.png) |
| Remora         | 1 oxygenerator             | 1000W       | 1200 O₂     | 13    | [<img alt=Remora width=250 src="https://user-images.githubusercontent.com/2483420/137971094-f7e7d11b-3760-461d-aee8-031948d8bdf3.png" />](https://barotraumagame.com/baro-wiki/images/4/46/Remora.png) |
| Typhon         | 1 oxygenerator             | 1000W       | 1000 O₂     | 20    | [<img alt=Typhon width=250 src="https://user-images.githubusercontent.com/2483420/137971125-2ccca2ee-4530-4c1d-bba9-6461177c9472.png" />](https://barotraumagame.com/baro-wiki/images/a/ae/Typhon.png) |
| Typhon2        | 1 shuttleoxygenerator[²]() | 1000W       |  350 O₂     | 15    | [<img alt=Typhon2 width=250 src="https://user-images.githubusercontent.com/2483420/137971152-2260fb10-02f7-46ef-8d15-540825bb86d2.png" />](https://barotraumagame.com/baro-wiki/images/a/a1/Typhon_2.png) |
| Hemulen        | 1 shuttleoxygenerator      | 1000W       |  800 O₂     |  5    | [<img alt=Hemulen width=250 src="https://barotraumagame.com/baro-wiki/images/thumb/6/6c/Hemulen.png/1500px-Hemulen.png" />](https://barotraumagame.com/baro-wiki/images/thumb/6/6c/Hemulen.png/1500px-Hemulen.png) |
| Selkie         | 1 shuttleoxygenerator      | 1000W       |  300 O₂     |  3    | [<img alt=Selkie width=250 src="https://barotraumagame.com/baro-wiki/images/thumb/9/90/Selkie.png/1500px-Selkie.png" />](https://barotraumagame.com/baro-wiki/images/thumb/9/90/Selkie.png/1500px-Selkie.png) |
| Venture        | 1 shuttleoxygenerator      | 1000W       |  600 O₂     |  6    | [<img alt=Venture width=250 src="https://barotraumagame.com/baro-wiki/images/thumb/f/fd/Venture.png/1500px-Venture.png" />](https://barotraumagame.com/baro-wiki/images/thumb/f/fd/Venture.png/1500px-Venture.png) |
| BeaconStation1 | 1 oxygenerator             | 1000W       |  400 O₂     |  6    | [<img alt=Beacon width=250 src="https://user-images.githubusercontent.com/2483420/129430270-13cd87a3-b624-4424-ab96-7e4ef176f036.png" />](https://user-images.githubusercontent.com/2483420/129430270-13cd87a3-b624-4424-ab96-7e4ef176f036.png) |
| Sub Editor     | 1 shuttleoxygenerator      | 1000W       |  400[³]() O₂|  -    | [<img alt=shuttleoxygenerator width=250 src="https://barotraumagame.com/baro-wiki/images/thumb/9/9b/Shuttle_Oxygen_Generator.png/280px-Shuttle_Oxygen_Generator.png" />](https://barotraumagame.com/baro-wiki/images/thumb/9/9b/Shuttle_Oxygen_Generator.png/280px-Shuttle_Oxygen_Generator.png) |
| Sub Editor     | 1 oxygenerator             | 1000W       |  400[³]() O₂|  -    | [<img alt=oxygenerator width=250 src="https://barotraumagame.com/baro-wiki/images/thumb/c/c3/Oxygen_Generator.png/280px-Oxygen_Generator.png" />](https://barotraumagame.com/baro-wiki/images/thumb/c/c3/Oxygen_Generator.png/280px-Oxygen_Generator.png) |
| Sub Editor     | 1 outpostoxygenerator      | 1000W       | 5000 O₂     |  -    | [<img alt=outpostoxygenerator width=250 src="https://user-images.githubusercontent.com/2483420/129430319-18b9e2da-9b7a-4aef-b6e8-5f91b1b2f9e6.png" />](https://user-images.githubusercontent.com/2483420/129430319-18b9e2da-9b7a-4aef-b6e8-5f91b1b2f9e6.png) |

[¹]() The top right vent has invalid link to top left guitar. 🐞

[²]() Why is this very large vessel using a _Shuttle_ Oxygen Generator? No mention of it in the ship lore/description.

[³]() Why are the shuttle and sub generators' amounts _the same_, and for the same power? Judging by the vanilla sub values, the sub generator should produce closer to 1000, and the shuttle should consume less power. However looking at it from a "recommended crew" perspective and 7 O₂ minimum per person, the base shuttle gen should produce something like 30 O₂ (4 ppl), the base sub gen 75 O₂ (10 ppl), and the outpost gen 400 O₂ (57 ppl).

<details>
  <summary>XML allowedlinks</summary>
  Except doors/hatches/JBs used for outposts.

  | item | allowedlinks |
  |-----:|:-------------|
  | navterminal | statusmonitor |
  | shuttlenavterminal | statusmonitor |
  | sonarmonitor | statusmonitor |
  | statusmonitor | navterminal |
  | fabricator | deconstructor,locker |
  | medicalfabricator | deconstructor,medcabinet,locker |
  | deconstructor | fabricator,medicalfabricator,locker |
  | (Legacy) navterminal | statusmonitor |
  | (Legacy) statusmonitor | command,navigationterminal |
  | (Legacy) oxygenerator | vent |
  | (Legacy) railgun | railgunammosource |
  | (Legacy) railgunloader | railgun |
  | (Legacy) railgunloadersinglevertical | railgun |
  | (Legacy) railgunloadersinglehorizontal | railgun |
  | (Legacy Forward) railgunloadersinglehorizontal | railgun |
  | (Legacy Rear) railgunloadersinglehorizontal | railgun |
  | oxygenerator | vent |
  | shuttleoxygenerator | vent |
  | vent | oxygenerator |
  | loosevent | oxygenerator |
  | railgunwrecked | railgunammosource |
  | railgunloaderwrecked | railgun |
  | railgunloadersingleverticalwrecked | railgun |
  | railgunloadersinglehorizontalwrecked | railgun |
  | coilgunwrecked | coilgunequipment |
  | coilgunloaderwrecked | coilgun |
  | oxygeneratorwrecked | vent |
  | shuttleoxygeneratorwrecked | vent |
  | ventwrecked | oxygenerator |
  | chaingun | coilgunequipment |
  | chaingunloader | chaingun |
  | coilgun | coilgunequipment |
  | coilgunloader | coilgun |
  | depthchargetube | depthchargeammosource |
  | depthchargeloader | depthchargelauncher |
  | pulselaser | pulselaserequipment |
  | pulselaserloader | pulselaser |
  | railgun | railgunammosource |
  | railgunloadersmall | railgun |
  | turrethardpoint | turretammosource |
  | blankloader | turrethardpoint |
  | outpostoxygenerator | vent |
  | fleshgun1 | fleshgunloader |
  | fleshgunloader1 | fleshgun |
</details>

**Comment by MatthiasKunnen** [comment]
I really like this idea. It would also make it possible to cycle through connected cabinets using the new demux component

**Comment by NotWendy** [comment]
Since you're wondering why the default oxygen production numbers are the way they are, it's because the oxygen is split between rooms, and 400 is a semi-arbitrarily chosen number which ensures that multiple people in any one room (as long as it's not tiny) won't start suffocating. Dropping it to 75 will make rooms unbreathable very quickly, and will force the crew to either keep the doors and hatches open at all times, or else constantly move between rooms. (Picture a crying captain here.)
The number could be lower than 400 on most vanilla submarines, which would make air quality more of a concern. But going below 200 makes it kind of a pain, especially when you consider that as the generator's condition deteriorates, it produces less and less oxygen per second.
Default power consumption is also a mostly meaningless stat, as what's important is how well the reactor/shuttle batteries can handle that load.
Typhon 2 using a shuttle oxygen generator is very funny, but that's just vanilla submarines being a mess.
Also, what the hell, this is a post from 2021. I didn't even realize that before I wrote the comment.

**Reply by MatthiasKunnen** [reply]
My reply will have pushed it back to the top of the activity pile. With some more interest we might see this implemented.

Also, picturing a captain running outside the helm room all the time to prevent suffocation was pretty funny 😄

**Comment by SomeRandomNoobKekeke** [comment]
Wdym you can't? https://steamcommunity.com/sharedfiles/filedetails/?id=2852315967

**Comment by girlyguppy** [comment]
it would be really cool if this worked via wifi, because then we would finally have status monitor for drones (https://github.com/FakeFishGames/Barotrauma/discussions/15045), and the ability to add teleporters via xml mods (linking from a shuttle's depth charge tube to a mainsub's depth charge loader which has a boarding pod type thing)

**Reply by girlyguppy** [reply]
I also really think that most of the utility for this isn't for changing links ingame, because i really don't want to do that, but the option existing in case I *wanted* to would be nice. it is however much better than them being linked via magic rather than an actual physical thing that can be interacted with. so most of the links i would leave locked such as ship guns to loaders, but something like an o2 generator to a vent, i could see the gameplay in being able to change around the links, especially with modded placeable vents and o2 generators.
